### ISSUE TITLE
Model evaluation

### ISSUE THREAD

**Issue by gulleh** [frgfm/torch-cam#14]
Hi,

I have a question about cam_example.py. Shouldn't the model be in eval() mode? I noticed without setting the model.eval() the prediction labels are wrong while after setting to eval mode the ScoreCam and SSCAM heatmaps are zero. Could you please check this?

**Comment by frgfm** [comment]
Hey there @gulleh,
I just made a PR #15 to address this. However, I was not able to reproduce your issue of having zero-value heatmaps for ScoreCAM and SSCAM. Did you use the default script arguments?

Here is what I got removing some CAMS:

![cam_example](https://user-images.githubusercontent.com/26927750/92329064-51070880-f065-11ea-9087-344db3a48eea.png)

**Comment by gulleh** [comment]
I made some changes to check the classification score first. Without setting to eval the predictions were wrong and the heat maps were are the same for different images. My problem is that after setting the model.eval(), the heatmaps won't change for different images.

**Comment by frgfm** [comment]
> I made some changes to check the classification score first. Without setting to eval the predictions were wrong and the heat maps were all the same for different images. My problem is that after setting the model.eval(), the heatmaps won't change for different images.

Thanks for the feedback! If you used different argument values than the default ones or modified the script, would you mind telling me? I'm unable to reproduce the behaviour you mentioned. As you can see, the output I got was from the script using the latest PR commit (only modifying the CAM selection), and the outputs are different from one method to another as shown on the picture I uploaded. 

Perhaps you were offline when you used the script? Then the pretrained parameters of the model would not have been loaded, and random initialization shields roughly similar classification scores.

Also you mentioned using it on different images, did you batch them ? or forward them one after another? (should work both ways, but there might be some issues with one of those methods). If you batched them, you have to change the `.squeeze(0)` for dispatching correctly the intended heatmaps (cf. https://github.com/frgfm/torch-cam/blob/master/scripts/cam_example.py#L70)
And for your image outputs, could you upload the result (with subplots showing the heatmaps that should be different)

Let me know if that helps!

**Comment by gulleh** [comment]
Thanks for the reply. Here is the snippet of the code I used.
```
fig, axes = plt.subplots(5, len(cam_extractors), figsize=(15, 15))
model.eval()
model.zero_grad()
for index in range(0, 5):
    img, target = iter(val_loader).next()
    for idx, extractor in enumerate(cam_extractors):
        scores = model(img.to(device))
        _, pred_label_idx = torch.topk(F.softmax(scores, dim=1), 1)
        predicted_label = idx_to_labels[str(pred_label_idx.item())][1]
        # Select the class index
        class_idx = scores.squeeze(0).argmax().item() if class_idx is None else class_idx
        if extractor is None:
            result = to_pil_image(denormalize(img.squeeze(0)))
        else:    
            # Use the hooked data to compute activation map
            activation_map = extractor(class_idx, scores).cpu()
            # Clean data
            extractor.clear_hooks()
            # Convert it to PIL image
            # The indexing below means first image in batch
            heatmap = to_pil_image(activation_map, mode='F')
            # Plot the result
            result = overlay_mask(to_pil_image(denormalize(img.squeeze(0))), heatmap)
        
        
        axes[index, idx].imshow(result)
        axes[index, idx].set_title(extractor.__class__.__name__, size=8)
        axes[index, idx].set_xlabel(predicted_label)
        target_label = idx_to_labels[str(target.item())][1]
        if predicted_label != target_label:
            print('Predicted:', predicted_label, 'Target is:', idx_to_labels[str(target.item())][1])
        axes[index,idx].axis('off')

plt.tight_layout()
plt.show()
```
![image](https://user-images.githubusercontent.com/1225441/92364149-97c03580-f12d-11ea-9aa5-feabf8678d2a.png)

As you can see the heatmap does not change for different images.

**Comment by frgfm** [comment]
Thanks for the details @gulleh!
I think I see why you get this behaviour: the sample script was designed to visualize different cams on a single picture, so I iterate only once per extractor. Which allows me to clear the hooks from the model before the next extractor gets to set his own.

In your case, on top of iterating on extractors, you iterate over pictures. When you keep the same extractor, but change the picture, if you use `.clear_hooks` it will basically clear all the hooks that were set when they were instantiated.

If I'm correct, you could easily correct this behaviour by:
- iterating first on extractors
- then iterating on pictures (without clearing hooks between them, only when you're finished with this CAM extractor)

Alternatively, if you want to keep your loop order, you would have to repopulate correctly the `hook_handles` attribute of the extractor, but it seems less efficient that my previous suggestion.

Let me know if you encounter other issues :)

**Comment by gulleh** [comment]
Thanks for the suggestion. However, I changed the position of clear_hooks and it seems to be working but I'm not sure.
![image](https://user-images.githubusercontent.com/1225441/92371674-dc9c9a00-f136-11ea-9541-47f75e067f92.png)
